### DISCUSSION TITLE
Clarification on EWW expressions

### DISCUSSION THREAD

**Discussion by DarkSamus9000** [elkowar/eww#435] [Q&A]
The [documentation for EWW expressions ](https://elkowar.github.io/eww/expression_language.html) is a little sparse, so I had some questions for you here if this is the place to ask them.

I have a widget with an optional attribute. Based on whether or not the attirubute is defined, I'd like the widget to include one of two different other widgets. But code speaks louder than words, so here's a relevant snippet of my `eww.yuck`:
```
(defwidget main [?second_monitor]
    (box :orientation "horizontal"
        :space-evenly "true"
        (group_left :second_monitor second_monitor)
        (group_center)
        (group_right)
    )
)

(defwidget group_left [?second_monitor]
    (box :orientation "horizontal"
        :class "group_left"
        :space-evenly "false"
        :halign "left"
        (dashboard)
        {second_monitor ? (workspaces_alt) : (workspaces)}
    )
)
```
EWW doesn't throw any errors, but this code doesn't produce what I'm expecting.
The widget `main` is produced, and I can see the `group_left` widget with its nested `dashboard` widget, but `group_left` is not including `workspaces_alt` or `workspaces`. It doesn't matter whether I define the attribute `second_monitor` or not when I create the `main` widget.

I suppose my primary question regards this line:
```
        {second_monitor ? (workspaces_alt) : (workspaces)}
```
Is this the correct way to do what I want? I want to check whether an attribute is defined or not, and depending on that, I want to include a different widget.

**Comment by elkowar** [comment]
optional arguments default to the value `""` IIRC, so the proper way to check if the default value is used is to do `second_monitor == ""`.
Next part: Eww expressions can't "return" other widgets, sadly.
What you want to do for this is do to something along the lines of this
```
(defwidget group_left [?second_monitor]
    (box :orientation "horizontal"
        :class "group_left"
        :space-evenly "false"
        :halign "left"
        (dashboard)
        (box :visible {second_monitor != ""} (workspaces_alt))
        (box :visible {second_monitor} (workspaces))
    )
)
```

then you can call it with
```
(group_left)
```
or
```
(group_left :second_monitor true)
```

**Reply by MartyMcFlyInTheSky** [reply]
What if I were to call it with `(group_left : second_monitor false)`, wouldn't this also take the `{second_monitor != ""}` branch?

**Comment by ItsTerm1n4l** [comment]
how would I write an if statement in eww then?

**Reply by Eclextic** [reply]
Create a [variable](https://elkowar.github.io/eww/configuration.html#adding-dynamic-content) and add a [ternary operator](https://elkowar.github.io/eww/expression_language.html) (in the docs its called a conditional) to dynamically adjust the widgets accordingly!

**Reply by ItsTerm1n4l** [reply]
Ok, thank you!

**Reply by ItsTerm1n4l** [reply]
I didn't realise it had to be done in a variable - probably why it didn't work for me.
